File: assignments/__lc/e_binary_search_bin_search704.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        end = len(nums) - 1
        middle = end // 2
        if nums[middle] == target:
            return middle
        elif nums[middle] < target:
            middle = middle + 1
            return helper(nums=nums, target=target, start=middle, stop=end)
        elif nums[middle] > target:
            middle = middle - 1
            return helper(nums=nums, target=target, start=0, stop=middle)


def helper(nums: List[int], target: int, start: int, stop: int) -> int:

    middle = (stop - start) // 2 + start
    print("start", start, "stop", stop, "middle", middle)
    if nums[middle] == target:
        print("MIDDLE")
        return middle
    elif start - stop == 1:
        return -1
    elif nums[middle] < target:
        middle = middle + 1
        return helper(nums=nums, target=target, start=middle, stop=stop)
    elif nums[middle] > target:
        middle = middle - 1
        return helper(nums=nums, target=target, start=start, stop=middle)


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            pivot = left + (right - left) // 2
            if nums[pivot] == target:
                return pivot
            if target < nums[pivot]:
                right = pivot - 1
            else:
                left = pivot + 1
        return -1


def search(nums: List[int], target: int) -> int:
    left = 0
    right = len(nums) - 1
    while left <= right:
        middle = (left + right) // 2
        if nums[middle] == target:
            return middle
        elif target < nums[middle]:
            right = middle - 1
        else:
            left = middle + 1
    return -1


alist = [
    [[-1, 0, 3, 5, 9, 12], 9],  # 4
    [[-1, 0, 3, 5, 9, 12], 2],  # -1
]

for x in alist:
    print(Solution().search(x[0], x[1]))
for x in alist:
    print(search(x[0], x[1]))

File: assignments/__lc/e_binary_search_bin_search704.rs
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let len_n: usize = nums.len();
        let mut lo: isize = 0;
        let mut hi: isize = len_n as isize - 1;
        while lo <= hi {
            let mid = lo + (hi - lo) / 2;
            if nums[mid as usize] == target {
                return mid as i32;
            }
            if nums[mid as usize] > target {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        -1
    }
}

impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let len_n: usize = nums.len();
        let mut low:isize=0;
        let mut high: isize = len_n as isize -1;
        let mut middle: isize = 0;
        middle = (low + high) /2 ;
        println!("low {:?} high {:?} middle {:?}", low, high, middle);
        
        while low <= high{
            middle = (low + high) /2 ;  
            if nums[middle as usize] == target{
                return middle as i32
            } else if nums[middle as usize] > target{
                high = middle -1
            } else{
                low = middle +1
            }
        }
    -1
    }
}